### PR TITLE
Display number of elements specified in "limit" field

### PR DESCRIPTION
**Pull request by mushketyk**
Instead of displaying up to 100 records now displays up the number of records provided in "limit" field.

### CODE REVIEW

**Comment by marcesher** [comment]
thanks @mushketyk 

This one still needs some work. The current 100 cap is there for a reason: it prevents users from crashing their browsers. For example, with your fix, try using a limit of 200000... it'll cause your browser to either take a very long time to load, or crash.

However, while the current behavior is safe, it's not quite correct. It does tell you that it's limiting the results to 100, but it still lets you choose more than that, and the pagination bar is incorrect.

We will retain the current 100-row cap for the HTML view, but if you wish to submit a fix for the number displayed, we'd be happy to have that. Thanks!

**Comment by mushketyk** [comment]
So what should be the correct behaviour? Should we just validate that offset is <= 100 and display an error message if a greater value is specified?

**Comment by marcesher** [comment]
@mushketyk Good question. I think that behavior makes sense.

Thanks!

**Comment by mushketyk** [comment]
I've updated the behaviour and added validation for the "limit" field.

**Comment by cndreisbach** [comment]
I think limit should be able to be higher than 100 when you're asking for anything but HTML. Is that right, @marcesher?

**Comment by marcesher** [comment]
@cndreisbach Yes, the limit applies only to the HTML view. All other formats are unconstrained (for better or worse)

**Comment by mushketyk** [comment]
I've updated the PR to validate max limit only for HTML viewer.
I am not sure about my design though. Initially I wanted to put representation to Query and validate everything together, but than I thought that representation and query parameters are different concerns.

**Comment by marcesher** [comment]
@mushketyk Thanks a lot. I'm going to try to make time to review this this week.

File: test/qu/test/query/validation.clj [hunk 1]
```diff
@@ -66,10 +67,14 @@
       (does-contain (errors (q :limit "ten")) :limit)
       (does-not-contain (errors (q :limit "10")) :limit))
 
+    (testing "it errors if limit is greater than 100"
+      (does-contain (limit-errors (q :limit "101")) :limit)
+      (does-not-contain (limit-errors (q :limit "100")) :limit))
+
     (testing "it does not error if limit is greater than 1000"
```
Comment (marcesher) [inline test/qu/test/query/validation.clj]: This is confusing... the new test checks for errors if the limit is greater than 100, but the existing test still checks for absence of error if limit is greater than 1000.

**Comment by marcesher** [comment]
@mushketyk @cndreisbach @m3brown I'm a little torn on this one, but I think it's pretty close.

Ivan's changes make the UI more technically correct... it avoids the pagination bug entirely, and provides an error message in a proper place. I like that.

But I think it needs a bit more... maybe it's just a matter of merging the existing message about "more results available in other formats" with the new validation error message. I've done that locally and am attaching a screenshot. It's not super pretty, but it keeps the message in one place and, again, it's more technically correct than the existing behavior.

As for Ivan's concerns about the implementation... @cndreisbach do you have any opinion on this? I don't have strong reservations about it. In the absence of an existing resource-specific validation scheme, and no current plans to implement one, this seems like a good step to me.

One thing I would like to see though: since this gets rid of the "has-more-data" necessity, @mushketyk, please rip that out of views.clj and slice.mustache.

**Comment by marcesher** [comment]
screen shot 2016-01-12 at 8 02 39 am

What the error looks like when I merge the new validation with the existing message a user gets when entering more than 100

**Comment by m3brown** [comment]
Thanks for digging into this @marcesher and @mushketyk, I like the solution you've come up with!

**Comment by mushketyk** [comment]
Hi,

Thank you for your review.

I've rebased this change, updated error message and removed "has-more-data?" variable.

**Comment by jonathanwcrane** [comment]
This is stellar work, @mushketyk .  Contributions like yours are a rare gift.  Thank you.

**Comment by mushketyk** [comment]
Thank you @jonathanwcrane it was a pleasure.

**Comment by marcesher** [comment]
@mushketyk Thanks again for this!

I should've caught this before: just realized the edge case that would cause us to want to keep that "has more data" thing... when a user enters "0".  For HTML, this should retain the same behavior as if the user entered the maximum number of records, i.e. only show 100. I would not have a problem with the validation changing to a check for an integer between 1 and 100 for HTML. The message could stay the same as you have it now, because in Qu "0" denotes "unlimited".

Would you mind adding that one additional validation to the HTML view? All other types can stay as is.

Thanks!

**Comment by mushketyk** [comment]
@marcesher Sorry, I was quite busy lately.
I will add the check.

**Comment by marcesher** [comment]
@mushketyk thanks!

**Comment by mushketyk** [comment]
>> For HTML, this should retain the same behavior as if the user entered the maximum number of records, i.e. only show 100

It seems to me that the current behavior. When I enter "0" into the limit field web interface displays first 100 records. Should I test something else?

I added an additional validation test that checks if 0 is valid "limit" value.

**Comment by marcesher** [comment]
@mushketyk Thanks! So... I think it's quite close. The old behavior that I think we should retain is the "Only first 100 results..." message.

screen shot 2016-02-23 at 2 34 08 pm

**Comment by sleitner** [comment]
Many thanks for your contribution @mushketyk. This is super helpful. We'll merge this and then issue a separate PR to restore the "Only first 100 results..." message that @marcesher mentioned.

**Pull request merged**